### PR TITLE
Challenge deploy

### PR DESCRIPTION
**Pull request by ashpect**
/deploy and /delete:challenge routes for challenge. Creates challenge folder, then challenge folder, copies zip, unzips , creates image, deploys the image along with its Nodeport service. Delete route deletes the deployment and service of the challenge across all namespaces.

### CODE REVIEW

File: services/master/controllers/deploy.go [hunk 1]
```diff
@@ -1,15 +1,161 @@
...
+	fmt.Println("Testing" + localFilePath + "....and..." + pathInPod)
+	//deployer.CopyInPod(localFilePath, pathInPod)
+
+}
+
+func buildimage(foldername string) {
+	// Build the challenge with Dockerfile
+	dirPath, _ := os.Getwd()
+	fmt.Println("Dockerfile for the image is at :")
+	fmt.Println(dirPath + "/chall/" + foldername + "/" + foldername)
```
[the remark this answers is not in the document]
Comment (ashpect) [reply services/master/controllers/deploy.go:29]: will do

File: services/master/controllers/deploy.go [hunk 2]
```diff
@@ -1,15 +1,161 @@
...
-	localFilePath := "/home/percy/notekeeper.tar.gz"
-	pathInPod := "/opt/katana/katana_web_notekeeper.tar.gz"
-	deployer.DeployToAll(localFilePath, pathInPod)
+	challengetype := "web"
+	foldername := ""
+	fmt.Println("Starting")
+	if form, err := c.MultipartForm(); err == nil {
+
+		//sort this
+		if token := form.Value["token"]; len(token) > 0 {
```
[the remark this answers is not in the document]
Comment (ashpect) [reply services/master/controllers/deploy.go]: It was something for testing, removed

File: lib/utils/deploy.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	g "github.com/sdslabs/katana/configs"
```
Comment (Bisht13) [inline lib/utils/deploy.go:8]: "github.com/sdslabs/katana/configs" imported but not used as g (typecheck)

File: services/challengedeployerservice/createService.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
+package challengedeployerservice
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+
+	g "github.com/sdslabs/katana/configs"
```
Comment (Bisht13) [inline services/challengedeployerservice/createService.go:8]: "github.com/sdslabs/katana/configs" imported but not used as g (typecheck)

File: services/challengedeployerservice/createService.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/sdslabs/katana/lib/utils"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CreateService(chall_name, team_name string) (string, error) {
```
Comment (Bisht13) [inline services/challengedeployerservice/createService.go:14]: Argument variable naming convention

File: services/challengedeployerservice/createService.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+	team_namespace := team_name + "-ns"
```
Comment (Bisht13) [inline services/challengedeployerservice/createService.go:16]: Naming convention

File: services/challengedeployerservice/createService.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	//Get all services
+
+	//Check if service already exists
+	services, err := serviceClient.List(context.TODO(), metav1.ListOptions{})
+	for _, service := range services.Items {
+		if service.Name == chall_name {
+			fmt.Println("Service already exists for the challenge " + chall_name + " in namespace " + team_namespace)
+			return "", nil
```
Comment (Bisht13) [inline services/challengedeployerservice/createService.go:51]: Return the URL

File: services/challengedeployerservice/deleteChall.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
+package challengedeployerservice
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	g "github.com/sdslabs/katana/configs"
```
Comment (Bisht13) [inline services/challengedeployerservice/deleteChall.go:8]: "github.com/sdslabs/katana/configs" imported but not used as g (typecheck)

File: services/challengedeployerservice/deleteChall.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/sdslabs/katana/lib/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func DeleteChallenge(chall_name string) error {
```
Comment (Bisht13) [inline services/challengedeployerservice/deleteChall.go:13]: Argument variable naming convention

File: services/challengedeployerservice/deleteChall.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+	totalteams := utils.GetTeamNumber()
```
Comment (Bisht13) [inline services/challengedeployerservice/deleteChall.go:15]: Naming Convention

File: services/challengedeployerservice/deleteChall.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("---------------Deleting challenge for team: ", team_namespace)
+		serviceClient := kubeclient.CoreV1().Services(team_namespace)
+		deploymentsClient := kubeclient.AppsV1().Deployments(team_namespace)
+
+		//Get deployment
+		deps, err := deploymentsClient.Get(context.TODO(), chall_name, metav1.GetOptions{})
```
Comment (Bisht13) [inline services/challengedeployerservice/deleteChall.go:32]: Better naming

File: services/challengedeployerservice/deleteChall.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+		//Check if service exists
+		services, err := serviceClient.List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			fmt.Println(" Error in getting services for" + chall_name + "in the namespace " + team_namespace)
+			continue
+			//panic(err)
+		}
+
+		flag := 0
```
Comment (Bisht13) [inline services/challengedeployerservice/deleteChall.go:61]: The entire implementation can be done without creating a flag

File: services/master/controllers/deploy.go [hunk 13]
```diff
@@ -1,15 +1,153 @@
...
+	"strconv"
+
 	"github.com/gofiber/fiber/v2"
+	archiver "github.com/mholt/archiver/v3"
+	g "github.com/sdslabs/katana/configs"
+	"github.com/sdslabs/katana/lib/utils"
 	deployer "github.com/sdslabs/katana/services/challengedeployerservice"
 )
 
+func challcopy(dirPath, challengename, challengetype string) {
```
Comment (Bisht13) [inline services/master/controllers/deploy.go:17]: Naming convention

File: services/master/controllers/deploy.go [hunk 14]
```diff
@@ -1,15 +1,153 @@
...
+
+	localFilePath := dirPath + "/" + challengename + ".tar.gz"
+	pathInPod := "/opt/katana/katana_" + challengetype + "_" + challengename + ".tar.gz"
+	fmt.Println("Testing" + localFilePath + "....and..." + pathInPod)
+	deployer.CopyInPod(localFilePath, pathInPod)
+
+}
+
+func buildimage(folderName string) {
```
Comment (Bisht13) [inline services/master/controllers/deploy.go:26]: Naming convention

File: services/master/controllers/deploy.go [hunk 15]
```diff
@@ -1,15 +1,153 @@
...
+	dirPath, _ := os.Getwd()
+	fmt.Println("Dockerfile for the image is at :")
+	fmt.Println(dirPath + "/challenges/" + folderName + "/" + folderName)
+	cmd := exec.Command("docker", "build", "-t", folderName, dirPath+"/chall/"+folderName+"/"+folderName)
+	cmd2 := exec.Command("minikube", "image", "load", folderName)
+	cmd.Run()
+	cmd2.Run()
+}
+
+func createfolder(challengename string) (message int, newDirPath string) {
```
Comment (Bisht13) [inline services/master/controllers/deploy.go:37]: Naming convention

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: services/master/controllers/challengeUpdate.go [hunk 16]
```diff
@@ -64,25 +66,34 @@ func ChallengeUpdate(c *fiber.Ctx) error {
 	}
 
 	log.Println("Pull successful for", teamName, ". Building image...")
-
-	// Build the challenge with Dockerfile
+	firstPatch, err := exec.Command("docker", "inspect", dir).Output()
 	cmd := exec.Command("docker", "build", "-t", dir, "./teams/"+dir)
```
Comment (Bisht13) [inline services/master/controllers/challengeUpdate.go:70]: Use BuildDockerImage() from utils

**Pull request merged**